File: 2023/08/ts/main.ts
import { run } from 'utils/run.ts';

/** If part 2 test input has completely different input, set this to `true`. */
export const HAS_ALTERNATE = true;

export function part1(input: string): string {
   const sides = input.split('\n\n');
   const instructions = sides[0].replaceAll('L', '0').replaceAll('R', '1');
   const maps = sides[1]
      .split('\n')
      .map((str) => {
         const [dest, lr] = str.split(' = ');
         return [dest, lr.slice(1, lr.length - 1).split(', ')] as [string, [string, string]];
      })
      .reduce((pv, v) => {
         pv[v[0]] = v[1];
         return pv;
      }, {} as Record<string, [string, string]>);

   let i = 0;
   let nav = 'AAA';
   while (true) {
      const map = maps[nav];
      nav = map[instructions[i % instructions.length] as '0'];
      i++;
      if (nav === 'ZZZ') break;
   }
   return i.toString();
}

function gcd(a: number, b: number): number {
   return !b ? a : gcd(b, a % b);
}

function lcm(a: number, b: number): number {
   return (a * b) / gcd(a, b);
}

export function part2(input: string): string {
   const sides = input.split('\n\n');
   const instructions = sides[0].replaceAll('L', '0').replaceAll('R', '1');
   const maps = sides[1]
      .split('\n')
      .map((str) => {
         const [dest, lr] = str.split(' = ');
         return [dest, lr.slice(1, lr.length - 1).split(', ')] as [string, [string, string]];
      })
      .reduce((pv, v) => {
         pv[v[0]] = v[1];
         return pv;
      }, {} as Record<string, [string, string]>);

   const navs = Object.keys(maps).filter((s) => s[2] === 'A');
   const dists = new Array(navs.length).fill(0);
   for (let i = 0; i < navs.length; i++) {
      let j = 0;
      while (true) {
         const map = maps[navs[i]];
         navs[i] = map[instructions[j % instructions.length] as '0'];
         j++;
         if (navs[i][2] === 'Z') break;
      }
      dists[i] = j;
   }
   return dists.reduce((pv, v)=> lcm(pv, v), 1);
}

if (import.meta.main) {
   run(import.meta.url, part1, part2, HAS_ALTERNATE);
}

File: 2023/08/lua/main.lua
local run = require('src.lua.run')
local str = require "src.lua.utils.str"

local options = {
   has_alternate = false,
   has_io = false,
}

local function gcd(a, b)
   if b == 0 then
      return a
   end
   return gcd(b, a % b)
end

local function lcm(a, b)
   return (a * b) / gcd(a, b);
end

---@param input string
---@param is_test boolean
---@return string
local function part_1(input, is_test)
   local lines = str.split(input, '\n')
   local instructions = {}

   for i = 1, #lines[1] do
      local c = string.sub(lines[1], i, i)
      if c == "L" then
         instructions[#instructions + 1] = 0
      else
         instructions[#instructions + 1] = 1
      end
   end

   local maps = {}
   for i = 2, #lines do
      local chunks = str.split(lines[i], '=')
      local dest = string.sub(chunks[1], 1, 3)
      local lr = str.split(string.sub(chunks[2], 3, #chunks[2] - 1), ',')
      maps[dest] = { lr[1], string.sub(lr[2], 2) }
   end

   local i = 0
   local nav = "AAA"
   while true do
      local m = maps[nav]
      nav = m[instructions[(i % #instructions) + 1] + 1]
      i = i + 1
      if nav == "ZZZ" then
         break
      end
   end

   return tostring(i)
end

---@param input string
---@param is_test boolean
---@return string
local function part_2(input, is_test)
   local lines = str.split(input, '\n')
   local instructions = {}

   for i = 1, #lines[1] do
      local c = string.sub(lines[1], i, i)
      if c == "L" then
         instructions[#instructions + 1] = 0
      else
         instructions[#instructions + 1] = 1
      end
   end

   local maps = {}
   local navs = {}
   for i = 2, #lines do
      local chunks = str.split(lines[i], '=')
      local dest = string.sub(chunks[1], 1, 3)
      local lr = str.split(string.sub(chunks[2], 3, #chunks[2] - 1), ',')
      maps[dest] = { lr[1], string.sub(lr[2], 2) }

      if string.sub(dest, 3, 3) == "A" then
         navs[#navs + 1] = dest
      end
   end

   local res = 1
   for j = 1, #navs do
      local i = 0
      while true do
         local m = maps[navs[j]]
         navs[j] = m[instructions[(i % #instructions) + 1] + 1]
         i = i + 1
         if string.sub(navs[j], 3, 3) == "Z" then
            break
         end
      end
      res = lcm(res, i)
   end

   return tostring(res)
end

local function main()
   run(arg, part_1, part_2, options)
end

main()
